#### Project: CropGuard AI

The project uses two supervised ML tasks.

1. 🌾 Pre-Cultivation Crop Yield Prediction
ML Task: Regression
Target (Y): yield — existing numerical column
Purpose: Predict the expected crop yield in tonnes/hectare for a selected district, crop, season, and agricultural year using historical crop performance, rainfall, irrigation, and contextual information.
Output: Predicted Yield
Example: 4.82 tonnes/hectare
2. ⚠️ Agricultural Risk Classification
ML Task: Classification
Target (Y): risk_level — derived target
Classes: Low Risk / Moderate Risk / High Risk
Purpose: Identify the agricultural risk associated with a particular district–crop–season combination using historical crop performance, rainfall conditions, irrigation conditions, and other validated features.
Output: Risk Level + Risk Probability
Example: HIGH RISK — 78% probability

## Coding 

In [1]:
## importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import os 
path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
print(os.path.exists(path))

True


In [3]:
df = pd.read_csv(path)
print("Shape:", df.shape)


Shape: (10708, 29)


### 📤 0. Data Loading (from db load tables data as pandas df)


In [4]:
%pip install pymysql

Note: you may need to restart the kernel to use updated packages.


In [5]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus

username = "root"
password = input("Enter MySQL password: ")

engine = create_engine(
    f"mysql+pymysql://{username}:{quote_plus(password)}@localhost:3306/cropguard_ai"
)

with engine.connect() as connection:
    print("Database connection successful!")

Enter MySQL password:  Raju8008


Database connection successful!


In [6]:
import os 
path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
print(os.path.exists(path))

True


In [7]:
df.to_sql(
    "validated_agriculture_data",
    con=engine,
    if_exists="replace",
    index=False
)

print("Data imported successfully!")

Data imported successfully!


check = pd.read_sql(
    "SELECT COUNT(*) AS total_rows FROM validated_agriculture_data",
    engine
)

print(check)

In [8]:
df = pd.read_sql(
    """
    SELECT *
    FROM validated_agriculture_data
    """,
    engine
)

print("Data loaded from MySQL successfully!")
print("Shape:", df.shape)
df.head()

Data loaded from MySQL successfully!
Shape: (10708, 29)


,id,year,state_name,state_code,district_name,district_code,crop_name,crop_code,crop_type,season,...,normal_rainfall,rainfall_deviation,agri_year,irrigation_food_category,irrigation_crop_type,irrigation_crop_name,irrigation_crop_code,irrigation_season,crop_irrigation_area,total_irrigated_area
0,385411,2019-2020,Telangana,36,Adilabad,501,Bajra,103.0,Cereals,Kharif,...,1103.8,69.931180,2019-2020,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79570.0,1048794.0
1,385812,2019-2020,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.8,69.931180,2019-2020,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79570.0,1048794.0
2,407355,2020-2021,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.6,-15.100530,2020-2021,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79635.0,1101556.0
3,450641,2022-2023,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.6,68.352253,2022-2023,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79128.0,1040598.0
4,284709,2014-2015,Telangana,36,Adilabad,501,Gram,201.0,Pulses,Rabi,...,1103.6,89.294537,2014-2015,Food Crop,Pulses,Gram,201.0,Rabi,2019.0,271774.0


## REGRESSION TASK — CROP YIELD PREDICTION

## 1. Modeling 

- ## 1.1 Pre-Requisites

In [9]:
## y = df["yield"]

#### 1.1.1 X & y:
- Picking y output col for Predictive Modelling & considering remaining cols are X

In [10]:
## Target (y)
y = df["yield"]

In [11]:
## Features (X)
X = df[
    [
        "district_name",
        "crop_name",
        "crop_type",
        "season",
        "area",
        "actual_rainfall",
        "normal_rainfall",
        "rainfall_deviation",
        "total_irrigated_area"
    ]
]

#### 1.1.2 Train-Test Split:¶
Dividing X & y data into train-test parts , where train part will be used for model learning , test part will be used for trained model testing
Thumb rules: 75-25, 80-20, 70-30

Since this project is intended for pre-cultivation prediction, the data is
split chronologically rather than randomly.
Older agricultural years are used for model training and the most recent
years are kept for testing. This gives a more realistic estimate of how the
model may perform on future agricultural seasons.

In [12]:
## starting year
df["year_start"] = (
    df["agri_year"]
    .astype(str)
    .str.extract(r"(\d{4})")[0]
    .astype(int)
)


In [13]:
df = df.sort_values("year_start").reset_index(drop=True)
years = sorted(df["year_start"].unique())
print("Available agricultural years:")
print(years)

Available agricultural years:
[np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022)]


In [14]:
year_counts = df["year_start"].value_counts().sort_index()
print("Records available for each year:")
display(year_counts)

Records available for each year:


year_start
2014     455
2015     460
2016    1163
2017    1081
2018    1116
2019    1352
2020    1585
2021    1843
2022    1653
Name: count, dtype: int64

In [15]:
## chronological split 
split_index = int(len(years) * 0.80)

train_years = years[:split_index]
test_years = years[split_index:]

print("Training years:")
print(train_years)

print("\nTesting years:")
print(test_years)

Training years:
[np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020)]

Testing years:
[np.int64(2021), np.int64(2022)]


In [16]:
train_mask = df["year_start"].isin(train_years)
test_mask = df["year_start"].isin(test_years)

In [17]:
## splitting x and y 
X_train = X.loc[train_mask].copy()
X_test = X.loc[test_mask].copy()

y_train = y.loc[train_mask].copy()
y_test = y.loc[test_mask].copy()

print("X_train shape:", X_train.shape)
print("X_test shape :", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape :", y_test.shape)

X_train shape: (7212, 9)
X_test shape : (3496, 9)
y_train shape: (7212,)
y_test shape : (3496,)


####  1.1.3 Missing Values & Outliers Handling

Missing values are checked separately for input and target variables.

Missing values in the target are removed because the model cannot learn from
a record without a known yield.

Input missing values will be handled later through the preprocessing pipeline.

Yield outliers are investigated rather than automatically removed because
extreme values may represent genuine agricultural differences between crops.

In [18]:
print("Missing values in X_train:")
print(X_train.isnull().sum())

Missing values in X_train:
district_name           0
crop_name               0
crop_type               0
season                  0
area                    0
actual_rainfall         0
normal_rainfall         0
rainfall_deviation      0
total_irrigated_area    0
dtype: int64


In [19]:
print("Missing values in X_test:")
print(X_test.isnull().sum())

Missing values in X_test:
district_name           0
crop_name               0
crop_type               0
season                  0
area                    0
actual_rainfall         0
normal_rainfall         0
rainfall_deviation      0
total_irrigated_area    0
dtype: int64


In [20]:
## outliers Identification and Handling 
print("Yield statistics:")
print(y_train.describe())

Yield statistics:
count     7212.000000
mean         8.775623
std        248.847076
min          0.000000
25%          0.849750
50%          1.667000
75%          3.706250
max      21105.000000
Name: yield, dtype: float64


In [21]:
## IQR Method 
Q1 = y_train.quantile(0.25)
Q3 = y_train.quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR
print("Lower limit:", lower_limit)
print("Upper limit:", upper_limit)

Lower limit: -3.4349999999999987
Upper limit: 7.990999999999999


In [22]:
## potential outliers 
outlier_mask = (
    (y_train < lower_limit) |
    (y_train > upper_limit)
)

print("Potential yield outliers:", outlier_mask.sum())

Potential yield outliers: 815


In [23]:
## Inspecting outliers 
outlier_rows = X_train.loc[outlier_mask].copy()

outlier_rows["yield"] = y_train.loc[outlier_mask]

display(
    outlier_rows[
        [
            "district_name",
            "crop_name",
            "crop_type",
            "season",
            "area",
            "actual_rainfall",
            "normal_rainfall",
            "rainfall_deviation",
            "total_irrigated_area",
            "yield"
        ]
    ].sort_values("yield", ascending=False).head(20)
)

,district_name,crop_name,crop_type,season,area,actual_rainfall,normal_rainfall,rainfall_deviation,total_irrigated_area,yield
2417,Khammam,Coconut,Oilseeds,Whole Year,437.0,935.57,1095.7,6.442842,703084.0,21105.000
623,Khammam,Sugarcane,Sugar,Kharif,3139.0,1779.75,1095.7,37.852206,1168520.0,112.601
965,Wanaparthy,Sugarcane,Sugar,Kharif,934.0,758.02,749.0,32.187020,1862580.0,109.751
476,Jagitial,Sugarcane,Sugar,Kharif,1087.0,1578.24,978.4,67.992573,1880169.0,109.446
3778,Sangareddy,Sugarcane,Sugar,Kharif,12942.0,1189.49,954.9,77.740320,1049204.0,105.000
4283,Yadadri Bhuvanagiri,Sugarcane,Sugar,Kharif,26.0,997.51,744.3,53.002706,2209986.0,104.000
237,Adilabad,Sugarcane,Sugar,Kharif,10.0,1417.70,1103.6,47.542025,1041832.0,104.000
238,Kumuram Bheem Asifabad,Sugarcane,Sugar,Kharif,2.0,1417.70,1103.6,47.542025,1041832.0,104.000
239,Mancherial,Sugarcane,Sugar,Kharif,2.0,1417.70,1103.6,47.542025,1041832.0,104.000
475,Rajanna Sircilla,Sugarcane,Sugar,Kharif,89.0,1262.90,978.4,31.093714,1690626.0,103.236


In [24]:
## outliers in Numerical X Features 
numerical_cols = [
    "area",
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

In [25]:
for col in numerical_cols:
    Q1 = X_train[col].quantile(0.25)
    Q3 = X_train[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    count = (
        (X_train[col] < lower) |
        (X_train[col] > upper)
    ).sum()

    print(f"{col}: {count} potential outliers")

area: 1370 potential outliers
actual_rainfall: 0 potential outliers
normal_rainfall: 0 potential outliers
rainfall_deviation: 49 potential outliers
total_irrigated_area: 0 potential outliers


### Outlier Decision

The IQR method identified potential outliers in yield, area and rainfall
deviation.

The identified observations are retained because extreme values can occur
naturally in agricultural data. In particular, yield and cultivated area can
vary substantially across different crops and districts.

No observations are removed only on the basis of the IQR rule. Model
performance will be evaluated with the original observations, and the effect
of extreme values will be considered during model evaluation.

#### 1.1.4 Feature Engineering of x for y

#### 1.1.4.1 Feature Generation

For the baseline regression model, the available dataset already contains
the required agricultural, rainfall and irrigation features.

Therefore, no additional feature generation is performed at this stage.
The selected features will be evaluated directly during model training.

#### 1.1.4.2 Feature Selection
The following features were selected based on their relevance to crop yield
prediction and their availability in the prepared dataset.

In [26]:
selected_features = [
    "district_name",
    "crop_name",
    "crop_type",
    "season",
    "area",
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

print("Selected features:")
for feature in selected_features:
 print(feature)

Selected features:
district_name
crop_name
crop_type
season
area
actual_rainfall
normal_rainfall
rainfall_deviation
total_irrigated_area


#### 1.1.4.3 Feature Type Identification

The selected input variables contain both categorical and numerical data.

Categorical variables will be encoded into numerical form because machine
learning algorithms require numerical input.

Numerical variables will be considered for scaling, especially for
distance-based algorithms such as KNN and SVR.

In [27]:
categorical_features = [
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]

numerical_features = [
    "area",
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

print("Categorical features:")
print(categorical_features)

print("\nNumerical features:")
print(numerical_features)

Categorical features:
['district_name', 'crop_name', 'crop_type', 'season']

Numerical features:
['area', 'actual_rainfall', 'normal_rainfall', 'rainfall_deviation', 'total_irrigated_area']


#### 1.1.4.4 Missing Value Handling

Missing numerical values will be replaced using the median calculated from
the training data.

Missing categorical values will be replaced using the most frequent value
from the training data.

This prevents information from the test dataset from influencing the
training process.

#### 1.1.4.5 Encoding 
one-hot encoding for categorical columns 

In [28]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("encoder", OneHotEncoder(handle_unknown="ignore"))
    ]
)

#### 1.1.4.6 Scaling
For numerical features

In [29]:
from sklearn.preprocessing import StandardScaler
numerical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler())
    ]
)

In [30]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numerical_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

print("Preprocessing pipeline created successfully.")

Preprocessing pipeline created successfully.


In [31]:
print(preprocessor)

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('scaler', StandardScaler())]),
                                 ['area', 'actual_rainfall', 'normal_rainfall',
                                  'rainfall_deviation',
                                  'total_irrigated_area']),
                                ('cat',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('encoder',
                                                  OneHotEncoder(handle_unknown='ignore'))]),
                                 ['district_name', 'crop_name', 'crop_type',
                                  'season'])])


### 1.2 Model + Define, Train & Study

#### 1.2.1 Import Regression Algorithms

Different regression algorithms are trained and compared to identify the
model that performs best for crop yield prediction.

The following algorithms are considered:

1. Linear Regression
2. Polynomial Regression
3. Lasso Regression
4. Ridge Regression
5. KNN Regression
6. Support Vector Regression
7. Decision Tree Regression
8. Random Forest Regression
9. XGBoost Regression

In [32]:
## impoerting models
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.preprocessing import PolynomialFeatures
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

#### 1.2.2 Define Regression Models

In [33]:
models = {
    "Linear Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", LinearRegression())
    ]),

    "Polynomial Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("poly", PolynomialFeatures(degree=2, include_bias=False)),
        ("model", LinearRegression())
    ]),

    "Lasso Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", Lasso(alpha=0.01, max_iter=10000))
    ]),

    "Ridge Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", Ridge(alpha=1.0))
    ]),

    "KNN Regression": Pipeline([
        ("preprocessor", preprocessor),
        ("model", KNeighborsRegressor(n_neighbors=5))
    ]),

    "SVR": Pipeline([
        ("preprocessor", preprocessor),
        ("model", SVR())
    ]),

    "Decision Tree": Pipeline([
        ("preprocessor", preprocessor),
        ("model", DecisionTreeRegressor(random_state=42))
    ]),

    "Random Forest": Pipeline([
        ("preprocessor", preprocessor),
        ("model", RandomForestRegressor(
            n_estimators=100,
            random_state=42,
            n_jobs=-1
        ))
    ])
}

In [34]:
print("Regression models selected:\n")
for name in models:
    print( name)

Regression models selected:

Linear Regression
Polynomial Regression
Lasso Regression
Ridge Regression
KNN Regression
SVR
Decision Tree
Random Forest


#### 1.2.4 Train Models

In [35]:
trained_models = {}

for name, model in models.items():

    print(f"Training {name}...")

    model.fit(X_train, y_train)

    trained_models[name] = model

    print(f"{name} training completed.")

Training Linear Regression...
Linear Regression training completed.
Training Polynomial Regression...
Polynomial Regression training completed.
Training Lasso Regression...
Lasso Regression training completed.
Training Ridge Regression...
Ridge Regression training completed.
Training KNN Regression...
KNN Regression training completed.
Training SVR...
SVR training completed.
Training Decision Tree...
Decision Tree training completed.
Training Random Forest...
Random Forest training completed.


### 1.3 Trained Model Predictions & Evaluations

#### 1.3.1 Predictions on Train and Test Data

Predictions are generated using all trained regression models on both
training and test datasets.

The training predictions are used to understand model fitting, while
test predictions are used to evaluate how well the model performs on
unseen data.

In [36]:
predictions = {}

for name, model in trained_models.items():

    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    predictions[name] = {
        "y_train_pred": y_train_pred,
        "y_test_pred": y_test_pred
    }

    print(f"{name} predictions completed.")

Linear Regression predictions completed.
Polynomial Regression predictions completed.
Lasso Regression predictions completed.
Ridge Regression predictions completed.
KNN Regression predictions completed.
SVR predictions completed.
Decision Tree predictions completed.
Random Forest predictions completed.


#### 1.3.2 Regression Evaluation Metrics

In [37]:
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

results = []
for name in predictions:

    y_train_pred = predictions[name]["y_train_pred"]
    y_test_pred = predictions[name]["y_test_pred"]

    train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
    test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

    train_r2 = r2_score(y_train, y_train_pred)
    test_r2 = r2_score(y_test, y_test_pred)

    results.append({
        "Models": name,
        "TrainLoss (RMSE)": train_rmse,
        "TestLoss (RMSE)": test_rmse,
        "TrainScore (R²)": train_r2,
        "TestScore (R²)": test_r2
    })

#### 1.3.3 Bias-Variance Trade-off

In [38]:
for result in results:

    train_rmse = result["TrainLoss (RMSE)"]
    test_rmse = result["TestLoss (RMSE)"]

    train_r2 = result["TrainScore (R²)"]
    test_r2 = result["TestScore (R²)"]

    rmse_gap = test_rmse - train_rmse
    r2_gap = train_r2 - test_r2

    if train_r2 < 0.50 and test_r2 < 0.50:
        fit = "Underfit"

    elif r2_gap > 0.15:
        fit = "Overfit"

    else:
        fit = "Goodfit"

    result["Bias-Variance (Fit)"] = fit

#### 1.3.4 Final Regression Evaluation Table

In [39]:
evaluation_table = pd.DataFrame(results)

evaluation_table[
    [
        "Models",
        "TrainLoss (RMSE)",
        "TestLoss (RMSE)",
        "TrainScore (R²)",
        "TestScore (R²)",
        "Bias-Variance (Fit)"
    ]
]

,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Linear Regression,3.322033e+00,2.868486,0.999822,0.999935,Goodfit
1,Polynomial Regression,2.372171e+00,168.982916,0.999909,0.775950,Overfit
2,Lasso Regression,3.489937e+00,3.088659,0.999803,0.999925,Goodfit
3,Ridge Regression,1.153029e+02,170.285731,0.785279,0.772482,Goodfit
4,KNN Regression,1.988259e+02,356.955214,0.361529,0.000262,Underfit
5,SVR,2.486731e+02,356.997440,0.001259,0.000026,Underfit
6,Decision Tree,1.344680e-17,3.125262,1.000000,0.999923,Goodfit
7,Random Forest,9.482900e+01,136.405439,0.854763,0.854010,Goodfit


In [40]:
evaluation_table = evaluation_table.round({
    "TrainLoss (RMSE)": 2,
    "TestLoss (RMSE)": 2,
    "TrainScore (R²)": 2,
    "TestScore (R²)": 2
})

evaluation_table

,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Linear Regression,3.32,2.87,1.00,1.00,Goodfit
1,Polynomial Regression,2.37,168.98,1.00,0.78,Overfit
2,Lasso Regression,3.49,3.09,1.00,1.00,Goodfit
3,Ridge Regression,115.30,170.29,0.79,0.77,Goodfit
4,KNN Regression,198.83,356.96,0.36,0.00,Underfit
5,SVR,248.67,357.00,0.00,0.00,Underfit
6,Decision Tree,0.00,3.13,1.00,1.00,Goodfit
7,Random Forest,94.83,136.41,0.85,0.85,Goodfit


#### 1.3.5 Pick the best Model

In [41]:
model_ranking = evaluation_table.sort_values(
    by="TestLoss (RMSE)",
    ascending=True
).reset_index(drop=True)

model_ranking

,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Linear Regression,3.32,2.87,1.00,1.00,Goodfit
1,Lasso Regression,3.49,3.09,1.00,1.00,Goodfit
2,Decision Tree,0.00,3.13,1.00,1.00,Goodfit
3,Random Forest,94.83,136.41,0.85,0.85,Goodfit
4,Polynomial Regression,2.37,168.98,1.00,0.78,Overfit
5,Ridge Regression,115.30,170.29,0.79,0.77,Goodfit
6,KNN Regression,198.83,356.96,0.36,0.00,Underfit
7,SVR,248.67,357.00,0.00,0.00,Underfit


In [42]:
best_model_name = model_ranking.loc[0, "Models"]
print("Best Regression Model:", best_model_name)

Best Regression Model: Linear Regression



Based on the evaluation results, Linear Regression was selected as the
best regression model for crop yield prediction.

The model achieved the lowest Test RMSE of 2.87 and a Test R² of 1.00.
Its training and testing performance were also consistent, indicating
a good fit without a significant bias-variance problem.

Lasso Regression and Decision Tree also achieved a Test R² of 1.00,
but Linear Regression produced the lowest Test RMSE among the evaluated
models.

#### 1.4 Hyperparameter Tuning

Hyperparameter tuning is performed to check whether the performance of
the regression models can be improved further.

Linear Regression does not have a major regularization hyperparameter
to tune. Therefore, Ridge and Lasso regression are tuned using different
values of the alpha parameter.

The tuned models are evaluated on the test data and compared with the
original Linear Regression model.

### 1.5 Saving Model & Real-Time Prediction

#### 1.5.1 Select Final Model

Based on the model evaluation results, Linear Regression was selected as
the final regression model for crop yield prediction.

The trained Linear Regression pipeline includes the preprocessing steps
for numerical and categorical features along with the regression model.

In [43]:
final_model = trained_models["Linear Regression"]

print("Final model:", type(final_model.named_steps["model"]).__name__)

Final model: LinearRegression


#### 1.5.2 Save the Model

In [44]:
import joblib
joblib.dump(final_model, "cropguard_yield_model.pkl")

print("Model saved successfully.")

Model saved successfully.


In [45]:
loaded_model = joblib.load("cropguard_yield_model.pkl")
print("Saved model loaded successfully.")

Saved model loaded successfully.


In [46]:
# REAL-TIME CROP YIELD PREDICTION
import ipywidgets as widgets
from IPython.display import display, clear_output

In [47]:
data_path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
df = pd.read_csv(data_path)

In [48]:
loaded_model = joblib.load("cropguard_yield_model.pkl")
print("Yield model loaded successfully.")

Yield model loaded successfully.


In [49]:
# CLEAN CATEGORICAL COLUMNS
for col in [
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]:

    df[col] = (
        df[col]
        .astype(str)
        .str.strip()
        .str.lower()
    )

In [50]:
##  DISTRICT DROPDOWN

districts = sorted(
    df["district_name"]
    .dropna()
    .unique()
    .tolist()
)


district_dropdown = widgets.Dropdown(
    options=districts,
    description="District:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)

In [51]:
## CROP DROPDOWN 
crop_dropdown = widgets.Dropdown(
    options=[],
    description="Crop:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)

In [52]:
# SEASON DROPDOWN
season_dropdown = widgets.Dropdown(
    options=[],
    description="Season:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)

In [53]:
# AREA INPUT

area_input = widgets.FloatText(
    value=1.0,
    description="Area:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)


In [54]:
# PREDICT BUTTON
predict_button = widgets.Button(
    description="Predict Yield",
    button_style="success",
    layout=widgets.Layout(width="170px")
)


In [55]:
# OUTPUT
output = widgets.Output(
    layout=widgets.Layout(
        width="480px",
        border="1px solid lightgray",
        padding="10px"
    )
)


In [63]:

# PREDICTION FUNCTION

def predict_yield(button):

    with output:

        clear_output()

        try:

            # ---------------------------------------------
            # Get user selections
            # ---------------------------------------------

            selected_district = district_dropdown.value

            selected_crop = crop_dropdown.value

            selected_season = season_dropdown.value

            area = float(area_input.value)


            # ---------------------------------------------
            # Validate area
            # ---------------------------------------------

            if area <= 0:

                print("⚠️ Invalid Area")
                print("----------------------------------------")
                print(
                    "Please enter a land area greater than 0."
                )

                return


            # ---------------------------------------------
            # Find matching records
            # ---------------------------------------------

            matching_data = df[
                (df["district_name"] == selected_district) &
                (df["crop_name"] == selected_crop) &
                (df["season"] == selected_season)
            ].copy()


            if matching_data.empty:

                print("⚠️ Unable to generate prediction.")
                print("----------------------------------------")
                print(
                    "No information is available for "
                    "the selected combination."
                )

                return


            # ---------------------------------------------
            # Select latest available record
            # ---------------------------------------------

            matching_data = matching_data.sort_values(
                "year"
            )

            selected_data = matching_data.iloc[-1]


            # ---------------------------------------------
            # Backend features
            # ---------------------------------------------

            crop_type = selected_data["crop_type"]

            actual_rainfall = selected_data[
                "actual_rainfall"
            ]

            normal_rainfall = selected_data[
                "normal_rainfall"
            ]

            rainfall_deviation = selected_data[
                "rainfall_deviation"
            ]

            total_irrigated_area = selected_data[
                "total_irrigated_area"
            ]


            # ---------------------------------------------
            # Model input
            # ---------------------------------------------

            sample_input = pd.DataFrame({

                "district_name": [selected_district],

                "crop_name": [selected_crop],

                "crop_type": [crop_type],

                "season": [selected_season],

                "area": [area],

                "actual_rainfall": [actual_rainfall],

                "normal_rainfall": [normal_rainfall],

                "rainfall_deviation": [rainfall_deviation],

                "total_irrigated_area": [
                    total_irrigated_area
                ]

            })


            # ---------------------------------------------
            # Predict
            # ---------------------------------------------

            predicted_yield = loaded_model.predict(
                sample_input
            )[0]


            # ---------------------------------------------
            # Final output
            # ---------------------------------------------

            print("=" * 50)
            print("       🌾 CROP YIELD PREDICTION")
            print("=" * 50)

            print(
                f"District        : "
                f"{selected_district.title()}"
            )

            print(
                f"Crop            : "
                f"{selected_crop.title()}"
            )

            print(
                f"Season          : "
                f"{selected_season.title()}"
            )

            print(
                f"Area            : "
                f"{area:.2f} Hectare"
            )

            print("----------------------------------------")

            print(
                f"Predicted Yield : "
                f"{predicted_yield:.2f} Tonnes/Hectare"
            )

            print("=" * 50)


        except Exception as e:

            print("=" * 50)
            print("⚠️ PREDICTION ERROR")
            print("=" * 50)

            print(str(e))

            print("=" * 50)


#  CONNECT BUTTON

predict_button.on_click(
    predict_yield
)


##  DISPLAY


display(
    district_dropdown,
    crop_dropdown,
    season_dropdown,
    area_input,
    predict_button,
    output
)

Dropdown(description='District:', index=17, layout=Layout(width='380px'), options=('adilabad', 'bhadradri koth…

Dropdown(description='Crop:', index=11, layout=Layout(width='380px'), options=('arhar/tur', 'bajra', 'banana',…

Dropdown(description='Season:', index=1, layout=Layout(width='380px'), options=('all seasons', 'kharif', 'rabi…

FloatText(value=6.0, description='Area:', layout=Layout(width='380px'), style=DescriptionStyle(description_wid…

Button(button_style='success', description='Predict Yield', layout=Layout(width='170px'), style=ButtonStyle())

Output(layout=Layout(border_bottom='1px solid lightgray', border_left='1px solid lightgray', border_right='1px…

In [64]:
# CROP GUARD AI
# REAL-TIME CROP YIELD PREDICTION
import ipywidgets as widgets
from IPython.display import display, clear_output


# LOAD DATASET

data_path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"

df = pd.read_csv(data_path)


# LOAD FINAL YIELD MODEL

loaded_model = joblib.load("cropguard_yield_model.pkl")

print("Yield model loaded successfully.")


#CLEAN CATEGORICAL COLUMNS

for col in [
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]:

    df[col] = (
        df[col]
        .astype(str)
        .str.strip()
        .str.lower()
    )


# DISTRICT DROPDOWN

districts = sorted(
    df["district_name"]
    .dropna()
    .unique()
    .tolist()
)


district_dropdown = widgets.Dropdown(
    options=districts,
    description="District:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)


# CROP DROPDOWN

crop_dropdown = widgets.Dropdown(
    options=[],
    description="Crop:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)


#  SEASON DROPDOWN

season_dropdown = widgets.Dropdown(
    options=[],
    description="Season:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)


#  AREA INPUT

area_input = widgets.FloatText(
    value=1.0,
    description="Area:",
    style={"description_width": "60px"},
    layout=widgets.Layout(width="380px")
)


# PREDICT BUTTON

predict_button = widgets.Button(
    description="Predict Yield",
    button_style="success",
    layout=widgets.Layout(width="170px")
)


# oUTPUT

output = widgets.Output(
    layout=widgets.Layout(
        width="480px",
        border="1px solid lightgray",
        padding="10px"
    )
)


#  UPDATE CROPS WHEN DISTRICT CHANGES

def update_crops(change):

    selected_district = change["new"]

    district_data = df[
        df["district_name"] == selected_district
    ]

    available_crops = sorted(
        district_data["crop_name"]
        .dropna()
        .unique()
        .tolist()
    )

    crop_dropdown.options = available_crops

    season_dropdown.options = []

    if available_crops:

        crop_dropdown.value = available_crops[0]


# UPDATE SEASONS WHEN CROP CHANGES

def update_seasons(change):

    selected_district = district_dropdown.value

    selected_crop = change["new"]

    crop_data = df[
        (df["district_name"] == selected_district) &
        (df["crop_name"] == selected_crop)
    ]

    available_seasons = sorted(
        crop_data["season"]
        .dropna()
        .unique()
        .tolist()
    )

    season_dropdown.options = available_seasons

    if available_seasons:

        season_dropdown.value = available_seasons[0]


# CONNECT DROPDOWNS

district_dropdown.observe(
    update_crops,
    names="value"
)

crop_dropdown.observe(
    update_seasons,
    names="value"
)


# INITIALIZE CROPS

first_district = district_dropdown.value

district_data = df[
    df["district_name"] == first_district
]

initial_crops = sorted(
    district_data["crop_name"]
    .dropna()
    .unique()
    .tolist()
)

crop_dropdown.options = initial_crops


# INITIALIZE SEASONS

if initial_crops:

    first_crop = initial_crops[0]

    crop_data = district_data[
        district_data["crop_name"] == first_crop
    ]

    initial_seasons = sorted(
        crop_data["season"]
        .dropna()
        .unique()
        .tolist()
    )

    season_dropdown.options = initial_seasons


# PREDICTION FUNCTION
def predict_yield(button):

    with output:

        clear_output()

        try:

            # ---------------------------------------------
            # Get user selections
            # ---------------------------------------------

            selected_district = district_dropdown.value

            selected_crop = crop_dropdown.value

            selected_season = season_dropdown.value

            area = float(area_input.value)


            # ---------------------------------------------
            # Validate area
            # ---------------------------------------------

            if area <= 0:

                print("⚠️ Invalid Area")
                print("----------------------------------------")
                print(
                    "Please enter a land area greater than 0."
                )

                return


            # ---------------------------------------------
            # Find matching records
            # ---------------------------------------------

            matching_data = df[
                (df["district_name"] == selected_district) &
                (df["crop_name"] == selected_crop) &
                (df["season"] == selected_season)
            ].copy()


            if matching_data.empty:

                print("⚠️ Unable to generate prediction.")
                print("----------------------------------------")
                print(
                    "No information is available for "
                    "the selected combination."
                )

                return


            # ---------------------------------------------
            # Select latest available record
            # ---------------------------------------------

            matching_data = matching_data.sort_values(
                "year"
            )

            selected_data = matching_data.iloc[-1]


            # ---------------------------------------------
            # Backend features
            # ---------------------------------------------

            crop_type = selected_data["crop_type"]

            actual_rainfall = selected_data[
                "actual_rainfall"
            ]

            normal_rainfall = selected_data[
                "normal_rainfall"
            ]

            rainfall_deviation = selected_data[
                "rainfall_deviation"
            ]

            total_irrigated_area = selected_data[
                "total_irrigated_area"
            ]


            # ---------------------------------------------
            # Model input
            # ---------------------------------------------

            sample_input = pd.DataFrame({

                "district_name": [selected_district],

                "crop_name": [selected_crop],

                "crop_type": [crop_type],

                "season": [selected_season],

                "area": [area],

                "actual_rainfall": [actual_rainfall],

                "normal_rainfall": [normal_rainfall],

                "rainfall_deviation": [rainfall_deviation],

                "total_irrigated_area": [
                    total_irrigated_area
                ]

            })


            # ---------------------------------------------
            # Predict
            # ---------------------------------------------

            predicted_yield = loaded_model.predict(
                sample_input
            )[0]


            # ---------------------------------------------
            # Final output
            # ---------------------------------------------

            print("=" * 50)
            print("       🌾 CROP YIELD PREDICTION")
            print("=" * 50)

            print(
                f"District        : "
                f"{selected_district.title()}"
            )

            print(
                f"Crop            : "
                f"{selected_crop.title()}"
            )

            print(
                f"Season          : "
                f"{selected_season.title()}"
            )

            print(
                f"Area            : "
                f"{area:.2f} Hectare"
            )

            print("----------------------------------------")

            print(
                f"Predicted Yield : "
                f"{predicted_yield:.2f} Tonnes/Hectare"
            )

            print("=" * 50)


        except Exception as e:

            print("=" * 50)
            print("⚠️ PREDICTION ERROR")
            print("=" * 50)

            print(str(e))

            print("=" * 50)


# CONNECT BUTTON

predict_button.on_click(
    predict_yield
)


# DISPLAY

display(
    district_dropdown,
    crop_dropdown,
    season_dropdown,
    area_input,
    predict_button,
    output
)

Yield model loaded successfully.


Dropdown(description='District:', layout=Layout(width='380px'), options=('adilabad', 'bhadradri kothagudem', '…

Dropdown(description='Crop:', layout=Layout(width='380px'), options=('arhar/tur', 'bajra', 'banana', 'cashewnu…

Dropdown(description='Season:', layout=Layout(width='380px'), options=('all seasons', 'kharif', 'rabi'), style…

FloatText(value=1.0, description='Area:', layout=Layout(width='380px'), style=DescriptionStyle(description_wid…

Button(button_style='success', description='Predict Yield', layout=Layout(width='170px'), style=ButtonStyle())

Output(layout=Layout(border_bottom='1px solid lightgray', border_left='1px solid lightgray', border_right='1px…